### PR TITLE
[stdlib] Convert existential collection === operator to method

### PR DESCRIPTION
**Pull request by natecook1000**
#### What's in this pull request?

The `===` operator for existential collections is only used for standard library testing. This removes it in favor of an underscored method.
#### Resolved bug number: n/a

---

Before merging this pull request to apple/swift repository:
- [ ] Test pull request on Swift continuous integration.

<details>
  <summary>

Triggering Swift CI</summary>

The swift-ci is triggered by writing a comment on this PR addressed to the GitHub user @swift-ci. Different tests will run depending on the specific comment that you use. The currently available comments are:

**Smoke Testing**

| Platform | Comment |
| --- | --- |
| All supported platforms | @swift-ci Please smoke test |
| All supported platforms | @swift-ci Please smoke test and merge |
| OS X platform | @swift-ci Please smoke test OS X platform |
| Linux platform | @swift-ci Please smoke test Linux platform |

A smoke test on macOS does the following:
1. Builds the compiler incrementally.
2. Builds the standard library only for macOS. Simulator standard libraries and
   device standard libraries are not built.
3. lldb is not built.
4. The test and validation-test targets are run only for macOS. The optimized
   version of these tests are not run.

A smoke test on Linux does the following:
1. Builds the compiler incrementally.
2. Builds the standard library incrementally.
3. lldb is build incrementally.
4. The swift test and validation-test targets are run. The optimized version of these
   tests are not run.
5. lldb is tested.

**Validation Testing**

| Platform | Comment |
| --- | --- |
| All supported platforms | @swift-ci Please test |
| All supported platforms | @swift-ci Please test and merge |
| OS X platform | @swift-ci Please test OS X platform |
| OS X platform | @swift-ci Please benchmark |
| Linux platform | @swift-ci Please test Linux platform |

**Lint Testing**

| Language | Comment |
| --- | --- |
| Python | @swift-ci Please Python lint |

Note: Only members of the Apple organization can trigger swift-ci.
</details>

### CODE REVIEW

File: stdlib/public/core/ExistentialCollection.swift.gyb [hunk 1]
```diff
@@ -718,27 +718,20 @@ public func < (lhs: AnyIndex, rhs: AnyIndex) -> Bool {
...
-}
 
-/// Returns `true` iff `lhs` and `rhs` store the same underlying collection.
-public func === <
-  L : AnyCollectionProtocol, R : AnyCollectionProtocol
->(lhs: L, rhs: R) -> Bool {
-  return lhs._underlyingCollectionID == rhs._underlyingCollectionID
+  /// Returns `true` iff `self` and `other` use the same `_box` to store their
+  /// underlying collections.
+  func _isKnownToBeEqual<C: AnyCollectionProtocol>(to other: C) -> Bool
```
Comment (dabrahams) [inline stdlib/public/core/ExistentialCollection.swift.gyb]: Let’s call this `_storesSameUnderlyingCollection`

File: stdlib/public/core/ExistentialCollection.swift.gyb [hunk 2]
```diff
@@ -718,27 +718,20 @@ public func < (lhs: AnyIndex, rhs: AnyIndex) -> Bool {
...
 }
 
-/// Returns `false` iff `lhs` and `rhs` store the same underlying collection.
-public func !== <
-  L : AnyCollectionProtocol, R : AnyCollectionProtocol
->(lhs: L, rhs: R) -> Bool {
-  return lhs._underlyingCollectionID != rhs._underlyingCollectionID
+extension AnyCollectionProtocol {
+  public func _isKnownToBeEqual<C: AnyCollectionProtocol>(to other: C) -> Bool {
```
Comment (dabrahams) [inline stdlib/public/core/ExistentialCollection.swift.gyb]: let’s make this `public // @testable`.

**Comment by dabrahams** [comment]
Just a couple of fixups and this is ready to go.  Thanks, Nate!

File: stdlib/public/core/ExistentialCollection.swift.gyb (hunk 2 above)
Comment (gribozavr) [inline stdlib/public/core/ExistentialCollection.swift.gyb]: If this method is only called from tests, why not move it into the test file?

Comment (gribozavr) [inline stdlib/public/core/ExistentialCollection.swift.gyb]: And now that `AnyCollectionProtocol` does not have any public APIs, maybe underscore it as well?

Comment (gribozavr) [inline stdlib/public/core/ExistentialCollection.swift.gyb]: Another idea is that we can remove `AnyCollectionProtocol` from the library altogether, and re-add it in the test file.  We can just `unsafeBitCast` the `AnyCollection` to `UInt`, that would be our identifier (we would be relying on the fact that `AnyCollection` is a 1-word data structure, but we probably will not ever change that).

Comment (dabrahams) [inline stdlib/public/core/ExistentialCollection.swift.gyb]: on Tue Jul 19 2016, Dmitri Gribenko <notifications-AT-github.com> wrote:

> Another idea is that we can remove `AnyCollectionProtocol` from the
> library altogether, and re-add it in the test file.  We can just
> `unsafeBitCast` the `AnyCollection` to `UInt`, that would be our
> identifier (we would be relying on the fact that `AnyCollection` is a
> 1-word data structure, but we probably will not ever change that).

I'd prefer not to rely on unsafeBitCasts anywhere, ever.  Let's find
another way.

-- 
Dave

Comment (natecook1000) [inline stdlib/public/core/ExistentialCollection.swift.gyb]: Missed these, sorry! 

@gribozavr If we move `AnyCollectionProtocol` out of the library, we wouldn't be able to test this case without making `_box` public:

```swift
let a = AnyCollection([1, 2, 3, 4, 5])
let b = AnyRandomAccessCollection(a)!
// a. _storesSameUnderlyingCollection(b) == true
```
Proposal: add an underscore to `AnyCollectionProtocol` and change `_underlyingCollectionID` to `_boxID`. Thumbs up/down?

Comment (natecook1000) [inline stdlib/public/core/ExistentialCollection.swift.gyb]: Oh, and moving `_storesSameUnderlyingCollection` to the test file makes perfect sense.

Comment (dabrahams) [inline stdlib/public/core/ExistentialCollection.swift.gyb]: on Wed Jul 20 2016, Nate Cook <notifications-AT-github.com> wrote:

> Proposal: add an underscore to `AnyCollectionProtocol` and change
> `_underlyingCollectionID` to `_boxID`. Thumbs up/down?

SGTM
-- 
Dave

**Comment by natecook1000** [comment]
@swift-ci Please test

**Comment by natecook1000** [comment]
@swift-ci Please test

**Comment by gribozavr** [comment]
@swift-ci Please smoke test and merge

**Comment by gottesmm** [comment]
@swift-ci Please smoke test and merge

**Comment by gottesmm** [comment]
I cleared the workspaces. If this is good, it should work now.

**Comment by gribozavr** [comment]
@swift-ci Please test and merge

**Pull request merged**